### ISSUE TITLE
[osh] lost "strict errors" with latest echo changes (regression)

### ISSUE THREAD

**Issue by bar-g** [oilshell/oil#1772]
After the overhaul of the improved `echo` behaviour (with https://github.com/oilshell/oil/issues/1634 and https://github.com/oilshell/oil/issues/1605) , if users now  enable `simple_echo` (it's actually not a simple "single-arg" echo anymore, but still waiting to be renamed to `strict_echo` as of https://github.com/oilshell/oil/issues/1117) they now do not get any "strict errors" anymore to give notice that changes are required.

For example, if some shell script uses `echo -en "\nHello"`, it will just print `-en \nHello`, without any strict error, or warning, that the script should probably be using `printf "\nHello"` for compatible, reliable code.

Such feedback would seem appropriate as long as using osh and not yet ysh proper.

**Comment by bar-g** [comment]
Hm, the new `echo`'s behaviour may have simply lost its compatibility with echo in other shells.

The solution could then maybe be to split this into two "stepping" options `strict_echo` (disallowing shell pitfalls including -n and -e) and `ysh_echo` (as a known syntax for ysh write options, printing -n and -e as verbatim).

**Comment by andychu** [comment]
Yes the `echo` behavior arguably broke one of our language design principles

https://github.com/oilshell/oil/wiki/Language-Design-Principles

i.e. we try not to "take over" existing syntax, and instead provide new syntax

---

However I do think it's justified, although perhaps the FAQ doesn't justify it enough:

https://www.oilshell.org/release/0.19.0/doc/ysh-faq.html#how-do-i-write-echo-e-or-echo-n

---

First, there are some places where we silently changed syntax.  For example

```
ls $x
```

in YSH means something different than in OSH.  It doesn't do word splitting.  And it will also fail with `errexit`, even though the code looks the same.

Likewise

```
echo $x
```

means something different in YSH.  It doesn't respect `-e` if `x=-e`.  That's a GOOD thing IMO.


---

Also another justification is that you can usually do something like this:

```
egrep 'echo -[en][en]?' *.sh
```

And it will find all the usages you need to change.  I will add that to the doc

**Comment by andychu** [comment]
New docs - https://github.com/oilshell/oil/commit/182821424ad24c5676d5476a3b62b5ccac12a628

http://travis-ci.oilshell.org/github-jobs/5730/ovm-tarball.wwz/_release/VERSION/doc/ysh-faq.html#how-do-i-write-the-equivalent-of-echo-e-or-echo-n

**Comment by bar-g** [comment]
Oh, yes I fully agree it's better. (Just as in #1763 it's my use-case that puts the focus on strict errors.)

Could maybe osh and ysh:upgrade check and warn about possible options found in the string, while ysh proper is fine to always echo verbatim?

**Comment by bar-g** [comment]
Sorry, not osh, but strict_echo. (To be enabled individually, with strict_all and with ysh:upgrade, but not in ysh proper?)

**Comment by bar-g** [comment]
(Some users are always going to miss the FAQ, especially if not the same as user installing oils.)

**Comment by bar-g** [comment]
Wasn't there also some principle to avoid surprises?

I think the combination of strict_echo (parsing passed strings only for errors) and ysh_echo (current new behavior) may be able to avoid surprises without complicating usage (strict_echo being part of strict_all but not in ysh).

**Comment by bar-g** [comment]
Just to confirm explicitly, the (your) justification for the echo change itself seem ok. But there is still a problem with "loosing errors" (errors going unnoticed), the thing that took quite some effort to fix for errorexit. An FAQ item is just like a list of shell pitfalls just no true solution, unfortunately.

But it may be rather simple to fix this well using the scheme to combine two options like `parse_echoargs` and `strict_echoargs`:

**osh:**
Has `parse_echoargs` enabled by default.

With the optional `strict_echoargs` enabled, the parsing of options results in error messages hinting to the reliable shell and ysh alternatives.

**ysh:upgrade:**
enables `strict_echoargs`.

**ysh:**
Has both `parse_echoargs` and  `strict_echoargs` disabled by default.

But optionally enabling `strict_echoargs` could still allow to find all echoing of "option-like" strings (this could remain in FAQ item).

**Comment by andychu** [comment]
So `ysh:upgrade` and `ysh:all` are mostly runtime options, and all builtins like `echo`  are parsed at runtime too

I think the `egrep` solution I gave is actually the best one, because it's more static

It's true people may not find it until after they have a problem, but the same is true for anything like `strict_echoargs`

**Comment by andychu** [comment]
OK I see you're saying `strict_echoargs` can be on in the middle ...

Yeah I suppose that is possible

Right now we just respect it

```
$ bin/osh -o ysh:upgrade -c 'echo -n foo'
foo
```

OK so I guess it's not too hard to add a check for `-n` and `-e` there

That makes sense

**Comment by bar-g** [comment]
Yeah, great, I think that's it for properly fix-and-migrate echo then.

Because a `strict_echoargs` option,  separate from `parse_echoargs` (aka current `simple_echo`) allows for:

* An explicit strict_echoargs option available in osh (in addition to its parse_echoargs default).
* Automatic strict_echoargs by default for ysh:upgrade.
* Have all dyn parsing of echo params disabled in ysh by default (and also with ysh:all? or not?), i.e. strict_echoargs and parse_echoargs both off.
* An optional strict_echoargs option in ysh (implicitly enabling parse_echoargs?), to  check for forgotton usages of -e or -n.

**Comment by bar-g** [comment]
I guess better disable `strict_echoargs` in ysh:all, to keep it fully compatible to /bin/ysh.

For migrating code it's a bad idea to enable ysh:all directly (without first making the code work with ysh:upgrade), anyway.

**Comment by bar-g** [comment]
According to `shellcheck`, "In POSIX sh, echo flags are undefined. [SC3037]"

I think this lends to better names for the options: `parse_echoflags` (default on in osh) and `strict_echostrings` (defaults to off in osh). The latter already hinting to use the string languages! :-)
